### DISCUSSION TITLE
Writing an Ideal Heat Source component

### DISCUSSION THREAD

**Discussion by panosz** [oemof/tespy#415] [Q&A] [answered]
Hello. I want to create component for an idealized heat source. My intention is to use it in place of the combination of the heat source and pump in the `district heating basic example`, but without all the intricacies of the existing models. What I want to enforce is that the mass flow and fluid composition is the same between the inlet and the outlet. Other than this, any pressure ratio should be acceptable. Other parameters such as the HeatExchangerSimple.zeta should be ignored.  Can you help me get started?

**Comment by fwitte** [comment]
Hi @panosz,

thank you for reaching out! Could post a small sketch here, what the component should be able to do? What it’s parameters are, the equations, …? I am wondering, what the difference of a combination of cycle closer and pump is to what you are trying to do.

best

**Comment by panosz** [comment]
Hello @fwitte,

Thank you for you reply.
Since I am not very familiar with the subject, let me try to explain by example what I am trying to achieve.
My goal is to model small to moderate district heating networks. To do so I have been following the corresponding example in
[oemof-examples](https://github.com/oemof/oemof-examples#tespy). In that example, there is only one plant, that is represented by a pair of source
and sink, which is a reasonable choice, since it is the distribution side and not the production side that is of interest. 
If I understand correctly, since there is only one such pair, the condition `m_source = - m_sink` is implicitly satisfied, 
as the rest of the network is closed.

What I would like to do is extend this example to a network with multiple plants, but keeping the same level of detail in the model.
What I am worried about is that in this case, if I represent the plants as pairs of source and sink, I will not be able
to enforce the condition `m_source_i = - m_sink_i`, for each pair separately.

So, I guess the what I am asking is if there is a simple way to enforce these conditions in the model, without having to
add extra unnecessary complications to it.

**Reply by fwitte** [reply]
Ah okay got it. I guess you need to replace one source/sink construct with a cycle closer and a simple heat exchanger (red). And all other source/sink combinations with just a simple heat exchanger (green). There is only a single consumer in the graphics below (blue), but that should still make it work as intended.

I am not entirely sure though. So, I'd recommend you build a minimal example (just like in the sketch) and post the code here, if it does not work.

![grafik](https://github.com/oemof/tespy/assets/20252362/7432b6de-10b2-4786-9bb6-20baf02fcbe5)

Have a good day!

**Comment by panosz** [comment]
@fwitte,

Again, many thanks for your prompt reply.

There are some things that I don't particularly like with this approach.
  1) It adds extra complications, by introducing modelling details (the intrenals of the heat exchanger) that were
  previously ignored. 

  2) Usually, I get some kind of warning about the pressure ratio of the heat exchangers being greater than 1. 
 3) I can overcome '2' by adding a pump after each Heat exchanger that models a plant, but then there is an extra
 connection for each plant, which I have to initialize successfully and this is not always trivial to do.

Overall, it seems like an overkill when all I need to do is enforce some extra mass conservation constraints.

**Reply by fwitte** [reply, marked as answer]
So having a source and a sink in combination has the same total amount of connections as with the heat exchanger simple approach. There should not be any difference, because the equations will also be the same. If you want to go with source/sink instead of the heat exchanger you can use a referencing constraint:

```python
from tespy.connections import Ref

[...]

c_to_sink = Connection(...)
c_from_source = Connection(...)

c_from_source.set_attr(m=Ref(c_to_sink, 1, 0))
```

Effectively that is what the heat exchanger simple does in the background. If you want to avoid the pressure warning message (it is only a waring message letting you know, there is something uncommon, so nothing to worry in your specific case) use

```python
heat_ex = HeatExchangerSimple('somelabel')
heat_ex.pr.max_val = 1e3  # something sufficiently big actually
```
Best

**Reply by fwitte** [reply]
Documentation on referencing connection parameters is found here:

- https://tespy.readthedocs.io/en/main/modules/connections.html#connections
- https://tespy.readthedocs.io/en/main/api/connections.html#tespy.connections.connection.Ref (it shoud be "Ref" instead of "Bus" in the description of the class, that is a typo)

**Comment by panosz** [comment]
@fwitte

Thank you very much for the support!
